tests/lib/openstack_query/test_query_methods.py

```diff
@@ -110,66 +112,81 @@ def test_where_with_kwargs(self):
             ("with no kwargs and no subset", None, None),
         ]
     )
-    def test_run(self, _, mock_from_subset, mock_kwargs):
+    def test_run_with_optional_params(self, _, mock_from_subset, mock_kwargs):
         """
-        Tests that run method works expectedly
+        Tests that run method works expectedly - with subset and/or meta_params kwargs
         method should get client_side and server_side filters and forward them to query runner object
-        """
-        mock_query_builder = MagicMock()
-        self.instance.builder = mock_query_builder
-
-        mock_query_runner = MagicMock()
-        self.instance.runner = mock_query_runner
-
-        mock_query_output = MagicMock()
-        self.instance.output = mock_query_output
-
-        mock_client_filter_func = MagicMock()
-        mock_query_builder.client_side_filter = mock_client_filter_func
 
-        mock_server_filters = NonCallableMock()
-        mock_query_builder.server_side_filters = mock_server_filters
-
-        mock_query_results = NonCallableMock()
-        mock_query_runner.run.return_value = mock_query_results
+        """
+        mock_client_filter_func = self.mock_builder.client_side_filter
+        mock_server_filters = self.mock_builder.server_side_filters
+        mock_parser_results = ["obj1", "obj2"]
+        self.mock_parser.run_parser.return_value = mock_parser_results
 
         if not mock_kwargs:
             mock_kwargs = {}
 
-        res = self.instance.run("test-account", mock_from_subset, **mock_kwargs)
-        mock_query_runner.run.assert_called_once_with(
+        _ = self.instance.run("test-account", mock_from_subset, **mock_kwargs)
+        self.mock_runner.run.assert_called_once_with(
             "test-account",
             mock_client_filter_func,
             mock_server_filters,
             mock_from_subset,
             **mock_kwargs
         )
-        mock_query_output.generate_output.assert_called_once_with(mock_query_results)
+        self.mock_parser.run_parser.assert_called_once_with(
+            self.mock_runner.run.return_value
+        )
+        self.mock_output.generate_output.assert_called_once_with(["obj1", "obj2"])
 
-        self.assertEqual(res, self.instance)
+    @parameterized.expand(
+        [
+            ("parser outputs grouped results", True),
+            ("parser outputs list results", False),
+        ]
+    )
+    def test_run_with_parsing(self, _, parse_output_as_dict):
```

Comment:
Ditto, these are two test cases and IMO the parameterized and if-elses match a duplicated test case since we only have two cases

Plus it's easier to reason with two tests called `test_run_with_grouped_parsing` and `test_run_with_list_results`

Reply:
actually the `test_run_with_list_results` isn't needed since its the default and already being tested for free in other tests